File: baek/baek_14891_gear.py
import sys
sys.stdin = open('14891.txt', 'r')

gears = [list(map(int, input())) for _ in range(4)]  # S=1  N=0
int_gears = []
for gear in gears:
    cnt, ans = 7, 0
    while cnt != -1:
        tmp = gear.pop(0)
        if tmp:
            ans += 2 ** cnt
        cnt -= 1
    int_gears.append(ans)


def set_j(i):
    if i % 2 != chk:
        return not dir
    else:
        return dir


def turn(i, x):
    if x:
        int_gears[i] = (int_gears[i] | ((int_gears[i] & 1) << 8)) >> 1
    else:
        int_gears[i] = ((int_gears[i] << 1) | (int_gears[i] >> 7)) & ~(1 << 8)


def spread(i, flag):
    while 0 <= i < 3:
        if not xor_list[i]:
            break

        if flag == 1:
            i += 1

        j = set_j(i)
        turn(i, j)

        if flag == -1:
            i -= 1


K = int(input())
for _ in range(K):
    # 맞물린 곳이 같은 지 먼저 확인
    xor_list = []
    for i in range(3):
        xor_list.append(bool(int_gears[i] & (1 << 5)) ^ bool(int_gears[i + 1] & (1 << 1)))

    num, dir = map(int, input().split())
    if dir == -1:
        dir = 0

    i = num - 1
    chk = i % 2

    turn(i, dir)
    spread(i-1, -1)
    spread(i, 1)

res = 0
for i in range(4):
    if int_gears[i] & (1 << 7) == (1 << 7):
        res += 2 ** i
print(res)

File: baek/baek_14891_gear.cpp
#include <iostream>
#include <vector>
using namespace std;

int gears[6], K, xor_list[3];

void init() {
	freopen("./input.txt", "r", stdin);
	// vector<char> bits[9];
	char bits[9];

	for (int i = 1; i <= 4; i++) {
		cin >> bits;
		for (int j = 0; j < 8; j++) {
			// cout << bits[li] << " ";
			gears[i] <<= 1;
			gears[i] |= (bits[j] - '0');
		}
	}

	//for (int li = 0; li < 5; li++) {
	//	printf("%d \n", gears[li]);
	//}
	cin >> K;
}

void turn(int n, int d) {
	if (d > 0) {
		gears[n] = (gears[n] | ((gears[n] & 1) << 8)) >> 1;
	}
	else {
		gears[n] = ((gears[n] << 1) | ((gears[n] >> 7) & ~(1 << 8)));
	}

}

int set_j(int i, int dir, int chk) {
	if (i % 2 != chk) {
		return dir * -1;
	}
	else {
		return dir;
	}

}

void spread(int i, int d, int c, int flag) {
	while (0 <= i < 3) {
		if (!xor_list[i]) {
			break;
		}
		if (flag) {
			i += 1;
		}

		int j = set_j(i, d, c);
		turn(i, j);

		if (flag == -1) {
			i -= 1;
		}
	}
}

int sol() {
	int res = 0, num, dir, chk;

	while (K--) {
		for (int i = 0; i < 3; i++) {
			xor_list[i] = bool(gears[i] & (1 << 5)) ^ bool(gears[i+1] & (1 << 1));
		}
		//for (int li = 0; li < 3; li++) {
		//	printf("%d \n", xor_list[li]);
		//}

		cin >> num >> dir;
		num -= 1;
		chk = num % 2;

		turn(num, dir);
		spread(num - 1, dir, chk, -1);
		spread(num, dir, chk, 1);
	}

	for (int i = 0; i < 4; i++) {
		int tmp = 0;
		if (gears[i] & ((1 << 7) == (1 << 7))) {
			tmp = 1;
			for (int j = 0; j < i; j++) {
				tmp *= 2;
				//res += 2 ** i;
			}
			res += tmp;
		}
	}
	printf("%d", res);
	return res;
}

void main() {
	init();
	sol();
}